source/MRMesh/MRPolylineTrimWithPlane.cpp

```diff
@@ -5,175 +5,217 @@
 
 namespace MR
 {
-    EdgeBitSet subdividePolylineWithPlane( Polyline3& polyline, const Plane3f& plane, std::function<void( EdgeId, EdgeId, float )> onEdgeSplitCallback )
-    {
-        if ( polyline.topology.numValidVerts() == 0 )
-            return {};
 
-        EdgeBitSet res;
-        const auto sectionPoints = extractSectionsFromPolyline( polyline, plane, 0.0f );
-        for ( const auto& sectionPoint : sectionPoints )
-        {
-            const auto eNew = polyline.splitEdge( sectionPoint.e, polyline.edgePoint( sectionPoint.edgePointA() ) );            
-            res.autoResizeSet( sectionPoint.e );
-            if ( onEdgeSplitCallback )
-                onEdgeSplitCallback( sectionPoint.e, eNew, sectionPoint.a );
-        }
+EdgeBitSet subdivideWithPlane( Polyline3& polyline, const Plane3f& plane, EdgeBitSet* newPositiveEdges, std::function<void( EdgeId, EdgeId, float )> onEdgeSplitCallback )
+{
+    if ( polyline.topology.numValidVerts() == 0 )
+        return {};
 
-        return res;
-    }
+    if ( newPositiveEdges )
+        *newPositiveEdges = {};
 
-    UndirectedEdgeBitSet fillPolylineLeft( const Polyline3& polyline, const EdgeBitSet& orgEdges, std::vector<VertPair>* cutSegments )
+    EdgeBitSet sectionEdges;
+    const auto sectionPoints = extractSectionsFromPolyline( polyline, plane, 0.0f );
+    for ( const auto& sectionPoint : sectionPoints )
     {
-        const size_t numEdges = polyline.topology.lastNotLoneEdge().undirected() + 1;
-        UndirectedEdgeBitSet res( numEdges );
-        UndirectedEdgeBitSet visited( numEdges );
+        const auto eNew = polyline.splitEdge( sectionPoint.e, polyline.edgePoint( sectionPoint.edgePointA() ) );
+        sectionEdges.autoResizeSet( sectionPoint.e );
+        if ( onEdgeSplitCallback )
+            onEdgeSplitCallback( sectionPoint.e, eNew, sectionPoint.a );
+    }
 
-        if ( cutSegments )
-            cutSegments->reserve( orgEdges.count() / 2 );
+    EdgeBitSet visited = sectionEdges;
+    for ( EdgeId e : sectionEdges )
+    {
+        if ( visited.test( e ) )
+            continue;
```

Comment:
since `visited=sectionEdges`, we always `continue` here.

Reply:
rworked